File: next/electron/main/updateDownloader.ts
import { createHash, timingSafeEqual } from 'node:crypto'
import { mkdir, open, rename, rm } from 'node:fs/promises'
import { join } from 'node:path'
import type { UpdateDownload } from '../../src/shared/contracts/update'

type DownloadResponse = Pick<Response, 'ok' | 'status' | 'body'>
type DownloadFetch = (url: string, init: RequestInit) => Promise<DownloadResponse>

export type DownloadProgress = {
  percent: number
  transferred: number
  total: number
}

export async function downloadUpdateFile(
  download: UpdateDownload,
  directory: string,
  onProgress: (progress: DownloadProgress) => void,
  fetcher: DownloadFetch = fetch
): Promise<string> {
  await mkdir(directory, { recursive: true })
  const destination = join(directory, download.fileName)
  const temporary = join(directory, `.${download.fileName}.${process.pid}.download`)
  const response = await fetcher(download.url, {
    headers: { Accept: 'application/octet-stream', 'User-Agent': 'MooTool-Next' },
    redirect: 'follow',
    signal: AbortSignal.timeout(30 * 60 * 1_000)
  })
  if (!response.ok) throw new Error(`Update download returned HTTP ${response.status}`)
  if (!response.body) throw new Error('Update download returned an empty response')

  const file = await open(temporary, 'w', 0o600)
  const hash = createHash('sha512')
  const reader = response.body.getReader()
  let transferred = 0
  let closed = false

  try {
    onProgress({ percent: 0, transferred, total: download.size })
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      await writeAll(file, value)
      hash.update(value)
      transferred += value.byteLength
      onProgress({
        percent: Number(Math.min(100, transferred / download.size * 100).toFixed(1)),
        transferred,
        total: download.size
      })
    }
    if (transferred !== download.size) {
      throw new Error(`Update download size mismatch: expected ${download.size}, received ${transferred}`)
    }
    const actualChecksum = hash.digest()
    const expectedChecksum = Buffer.from(download.sha512, 'base64')
    if (expectedChecksum.length !== actualChecksum.length || !timingSafeEqual(actualChecksum, expectedChecksum)) {
      throw new Error('Update download checksum mismatch')
    }
    await file.sync()
    await file.close()
    closed = true
    await rename(temporary, destination)
    return destination
  } catch (error) {
    if (!closed) await file.close().catch(() => undefined)
    await rm(temporary, { force: true }).catch(() => undefined)
    throw error
  } finally {
    reader.releaseLock()
  }
}

async function writeAll(file: Awaited<ReturnType<typeof open>>, value: Uint8Array): Promise<void> {
  let offset = 0
  while (offset < value.byteLength) {
    const { bytesWritten } = await file.write(value, offset, value.byteLength - offset)
    if (bytesWritten <= 0) throw new Error('Could not write the downloaded update')
    offset += bytesWritten
  }
}

File: next-flutter/lib/core/update/update_downloader.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:path/path.dart' as p;

import 'update_models.dart';

typedef UpdateBytesFetcher = Future<UpdateBytesResponse> Function(Uri url);

class UpdateBytesResponse {
  const UpdateBytesResponse({required this.statusCode, required this.body});
  final int statusCode;
  final Uint8List body;
  bool get ok => statusCode >= 200 && statusCode < 300;
}

class DownloadProgress {
  const DownloadProgress({
    required this.percent,
    required this.transferred,
    required this.total,
  });
  final double percent;
  final int transferred;
  final int total;
}

class UpdateDownloader {
  UpdateDownloader({
    required this.directory,
    this.fetcher = defaultUpdateBytesFetcher,
    this.opener = openLocalPath,
  });

  final Directory directory;
  final UpdateBytesFetcher fetcher;
  final Future<bool> Function(String path) opener;
  bool _cancelled = false;

  void cancel() => _cancelled = true;

  Future<File> download(
    UpdateDownload download, {
    void Function(DownloadProgress progress)? onProgress,
  }) async {
    _cancelled = false;
    if (!download.url.startsWith('https://')) {
      throw FormatException('Update download must use HTTPS');
    }
    await directory.create(recursive: true);
    final destination = File(p.join(directory.path, download.fileName));
    final temporary = File(p.join(
        directory.path, '.${download.fileName}.${pidOrZero()}.download'));
    onProgress?.call(
        DownloadProgress(percent: 0, transferred: 0, total: download.size));
    final response = await fetcher(Uri.parse(download.url));
    if (_cancelled) {
      throw const UpdateCancelled();
    }
    if (!response.ok) {
      throw FormatException(
          'Update download returned HTTP ${response.statusCode}');
    }
    if (response.body.length != download.size) {
      throw FormatException(
          'Update download size mismatch: expected ${download.size}, received ${response.body.length}');
    }
    final actual = base64Encode(sha512.convert(response.body).bytes);
    if (actual != download.sha512) {
      throw FormatException('Update download checksum mismatch');
    }
    await temporary.writeAsBytes(response.body, flush: true);
    if (await destination.exists()) await destination.delete();
    await temporary.rename(destination.path);
    onProgress?.call(DownloadProgress(
        percent: 100, transferred: download.size, total: download.size));
    return destination;
  }

  Future<bool> openInstaller(File file) async {
    if (!await file.exists()) {
      throw FormatException('Downloaded update is missing');
    }
    return opener(file.path);
  }
}

class UpdateCancelled implements Exception {
  const UpdateCancelled();
  @override
  String toString() => 'Update download cancelled';
}

int pidOrZero() {
  try {
    return pid;
  } catch (_) {
    return 0;
  }
}

Future<UpdateBytesResponse> defaultUpdateBytesFetcher(Uri url) async {
  final client = HttpClient();
  try {
    client.connectionTimeout = const Duration(minutes: 2);
    final request = await client.getUrl(url);
    request.headers.set(HttpHeaders.acceptHeader, 'application/octet-stream');
    request.headers.set(HttpHeaders.userAgentHeader, 'MooTool-Next-Flutter');
    final response = await request.close().timeout(const Duration(minutes: 30));
    final builder = BytesBuilder(copy: false);
    await for (final chunk in response) {
      builder.add(chunk);
    }
    return UpdateBytesResponse(
      statusCode: response.statusCode,
      body: builder.takeBytes(),
    );
  } finally {
    client.close(force: true);
  }
}

Future<bool> openLocalPath(String path) async {
  if (Platform.isMacOS) {
    final result = await Process.run('open', [path]);
    return result.exitCode == 0;
  }
  if (Platform.isWindows) {
    final result = await Process.run('cmd', ['/c', 'start', '', path]);
    return result.exitCode == 0;
  }
  final result = await Process.run('xdg-open', [path]);
  return result.exitCode == 0;
}
